// flow-runtime/lib/cyclic.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.inValidationCycle = inValidationCycle;
exports.startValidationCycle = startValidationCycle;
exports.endValidationCycle = endValidationCycle;
exports.inToStringCycle = inToStringCycle;
exports.startToStringCycle = startToStringCycle;
exports.endToStringCycle = endToStringCycle;


// Tracks whether we're in validation of cyclic objects.
var cyclicValidation = new WeakMap();
// Tracks whether we're toString() of cyclic objects.


var cyclicToString = new WeakSet();

function inValidationCycle(type, input) {
  var tracked = cyclicValidation.get(type);
  if (!tracked) {
    return false;
  } else {
    return tracked.has(input);
  }
}

function startValidationCycle(type, input) {
  var tracked = cyclicValidation.get(type);
  if (!tracked) {
    tracked = new WeakSet();
    cyclicValidation.set(type, tracked);
  }
  tracked.add(input);
}

function endValidationCycle(type, input) {
  var tracked = cyclicValidation.get(type);
  if (tracked) {
    tracked.delete(input);
  }
}

function inToStringCycle(type) {
  return cyclicToString.has(type);
}

function startToStringCycle(type) {
  cyclicToString.add(type);
}

function endToStringCycle(type) {
  cyclicToString.delete(type);
}
