// flow-runtime/lib/declarations/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExtendsDeclaration = exports.ClassDeclaration = exports.ModuleExportsDeclaration = exports.ModuleDeclaration = exports.VarDeclaration = exports.TypeDeclaration = exports.Declaration = undefined;

var _Declaration = require('./Declaration');

var _Declaration2 = _interopRequireDefault(_Declaration);

var _VarDeclaration = require('./VarDeclaration');

var _VarDeclaration2 = _interopRequireDefault(_VarDeclaration);

var _TypeDeclaration = require('./TypeDeclaration');

var _TypeDeclaration2 = _interopRequireDefault(_TypeDeclaration);

var _ModuleDeclaration = require('./ModuleDeclaration');

var _ModuleDeclaration2 = _interopRequireDefault(_ModuleDeclaration);

var _ModuleExportsDeclaration = require('./ModuleExportsDeclaration');

var _ModuleExportsDeclaration2 = _interopRequireDefault(_ModuleExportsDeclaration);

var _ClassDeclaration = require('./ClassDeclaration');

var _ClassDeclaration2 = _interopRequireDefault(_ClassDeclaration);

var _ExtendsDeclaration = require('./ExtendsDeclaration');

var _ExtendsDeclaration2 = _interopRequireDefault(_ExtendsDeclaration);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.Declaration = _Declaration2.default;
exports.TypeDeclaration = _TypeDeclaration2.default;
exports.VarDeclaration = _VarDeclaration2.default;
exports.ModuleDeclaration = _ModuleDeclaration2.default;
exports.ModuleExportsDeclaration = _ModuleExportsDeclaration2.default;
exports.ClassDeclaration = _ClassDeclaration2.default;
exports.ExtendsDeclaration = _ExtendsDeclaration2.default;
