// flow-runtime/lib/types/TypeParameter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _Type2 = require('./Type');

var _Type3 = _interopRequireDefault(_Type2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * # TypeParameter
 *
 * Type parameters allow polymorphic type safety.
 * The first time a type parameter is checked, it records the shape of its input,
 * this recorded shape is used to check all future inputs for this particular instance.
 */
var TypeParameter = function (_Type) {
  _inherits(TypeParameter, _Type);

  function TypeParameter() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, TypeParameter);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = TypeParameter.__proto__ || Object.getPrototypeOf(TypeParameter)).call.apply(_ref, [this].concat(args))), _this), _this.typeName = 'TypeParameter', _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(TypeParameter, [{
    key: 'collectErrors',
    value: function collectErrors(validation, path, input) {
      var recorded = this.recorded,
          bound = this.bound,
          context = this.context;


      if (recorded) {
        return recorded.collectErrors(validation, path, input);
      } else if (bound && bound.collectErrors(validation, path, input)) {
        return true;
      }

      this.recorded = context.typeOf(input);
      return false;
    }
  }, {
    key: 'accepts',
    value: function accepts(input) {
      var recorded = this.recorded,
          bound = this.bound,
          context = this.context;


      if (recorded) {
        return recorded.accepts(input);
      } else if (bound && !bound.accepts(input)) {
        return false;
      }
      this.recorded = context.typeOf(input);

      return true;
    }
  }, {
    key: 'acceptsType',
    value: function acceptsType(input) {
      var recorded = this.recorded,
          bound = this.bound;

      if (input instanceof TypeParameter) {
        // We don't need to check for `recorded` or `bound` fields
        // because the input has already been resolved.
        return true;
      } else if (recorded) {
        return recorded.acceptsType(input);
      } else if (bound) {
        return bound.acceptsType(input);
      } else {
        // A generic type parameter accepts any input.
        return true;
      }
    }

    /**
     * Get the inner type or value.
     */

  }, {
    key: 'unwrap',
    value: function unwrap() {
      var recorded = this.recorded,
          bound = this.bound;

      if (recorded) {
        return recorded.unwrap();
      } else if (bound) {
        return bound.unwrap();
      } else {
        return this;
      }
    }
  }, {
    key: 'toString',
    value: function toString(withBinding) {
      var id = this.id,
          bound = this.bound;

      if (withBinding && bound) {
        return id + ': ' + bound.toString();
      }
      return id;
    }
  }, {
    key: 'toJSON',
    value: function toJSON() {
      return {
        typeName: this.typeName,
        id: this.id,
        bound: this.bound,
        recorded: this.recorded
      };
    }
  }]);

  return TypeParameter;
}(_Type3.default);

exports.default = TypeParameter;
